aptos-move/framework/aptos-stdlib/sources/cryptography/groth16.move

```diff
@@ -0,0 +1,169 @@
+module aptos_std::groth16 {
+    #[test_only]
+    use aptos_std::groups::{BLS12_381_G1, BLS12_381_G2, deserialize_scalar, deserialize_element_uncompressed, BLS12_381_Gt};
+    use aptos_std::groups;
+
+    // Error codes
+    const E_NATIVE_FUN_NOT_AVAILABLE: u64 = 1;
+
+    /// A Groth16 verifying key.
+    struct VerifyingKey<phantom G1, phantom G2, phantom Gt> has drop {
+        alpha_g1: groups::Element<G1>,
+        beta_g2: groups::Element<G2>,
+        gamma_g2: groups::Element<G2>,
+        delta_g2: groups::Element<G2>,
+        gamma_abc_g1: vector<groups::Element<G1>>,
+    }
+
+    /// A Groth16 verifying key pre-processed for faster verification.
+    struct PreparedVerifyingKey<phantom G1, phantom G2, phantom Gt> has drop {
+        alpha_g1_beta_g2: groups::Element<Gt>,
+        gamma_g2_neg: groups::Element<G2>,
+        delta_g2_neg: groups::Element<G2>,
+        gamma_abc_g1: vector<groups::Element<G1>>,
+    }
+
+    /// A Groth16 proof.
+    struct Proof<phantom G1, phantom G2, phantom Gt> has drop {
+        a: groups::Element<G1>,
+        b: groups::Element<G2>,
+        c: groups::Element<G1>,
+    }
+
+    /// Create a new Groth16 verifying key.
+    public fun new_vk<G1,G2,Gt>(alpha_g1: groups::Element<G1>, beta_g2: groups::Element<G2>, gamma_g2: groups::Element<G2>, delta_g2: groups::Element<G2>, gamma_abc_g1: vector<groups::Element<G1>>): VerifyingKey<G1,G2,Gt> {
+        abort_if_feature_disabled();
```

Comment:
So you don't need to abort in this function because it does not call a native. We just abort in the functions that call natives.

I realize that this leaves the module half-ready for use.

Reply:
Fixed.